File: android/app/src/main/java/com/bluelink/android/usb/UsbState.kt
package com.bluelink.android.usb

import com.bluelink.android.domain.*

enum class UsbStage { OFF, WAITING, AUTHORIZATION, PERMISSION_DENIED, NEGOTIATING, READY,
    HIGH_SPEED, FULL_SPEED, FALLBACK, UNAVAILABLE, DRIVER_MISSING, POLICY_BLOCKED, UNSUPPORTED }

data class UsbSnapshot(val stage: UsbStage = UsbStage.OFF, val peerName: String = "",
    val peerId: String? = null, val permissionGranted: Boolean = false,
    val bluetoothAvailable: Boolean = false)

internal object UsbStatePolicy {
    fun isReady(enabled: Boolean, peerId: String?, sessions: List<ManagedSessionState>) = enabled && !peerId.isNullOrBlank() &&
        sessions.any { it.phase == ConnectionPhase.CONNECTED && it.usbFileReady && it.peerId.equals(peerId, true) }

    fun noticeStage(enabled: Boolean, peerId: String?, ready: Boolean, snapshot: UsbSnapshot): UsbStage? =
        if (!enabled || ready || peerId.isNullOrBlank() || !snapshot.peerId.equals(peerId, true)) null
        else snapshot.stage.takeIf { it in setOf(UsbStage.AUTHORIZATION, UsbStage.NEGOTIATING, UsbStage.FALLBACK, UsbStage.UNAVAILABLE,
            UsbStage.PERMISSION_DENIED, UsbStage.DRIVER_MISSING, UsbStage.POLICY_BLOCKED, UsbStage.UNSUPPORTED) }
    fun isBlueLink(manufacturer: String?, model: String?) = manufacturer == "BlueLink" && model == "BlueLink"
    fun bluetoothAvailable(sessions: List<ManagedSessionState>, peerId: String?) = !peerId.isNullOrBlank() &&
        sessions.any { it.transport == SessionTransport.BLUETOOTH && it.phase == ConnectionPhase.CONNECTED && it.peerId.equals(peerId, true) }
    fun disconnected(sessions: List<ManagedSessionState>, peerId: String?) =
        if (bluetoothAvailable(sessions, peerId)) UsbStage.FALLBACK else UsbStage.UNAVAILABLE
}

File: windows/BlueLink.App/Usb/UsbState.cs
using BlueLink.Localization;

namespace BlueLink.Usb;

public enum UsbStage { Off, Waiting, Authorization, PermissionDenied, Negotiating, Ready, HighSpeed, FullSpeed, Fallback, Unavailable, DriverMissing, PolicyBlocked, Unsupported }
public enum UsbLinkSpeed { Unknown, FullSpeed = 2, HighSpeed = 3 }
public sealed record UsbSnapshot(UsbStage Stage = UsbStage.Off, string PeerName = "", string DeviceId = "",
    string Driver = "", UsbLinkSpeed Speed = UsbLinkSpeed.Unknown, bool BluetoothAvailable = false, string? Error = null)
{
    // Filled only by an authenticated session, never by USB descriptors or ADB enumeration.
    public string? PeerId { get; init; }
    public string Channel => Stage switch
    {
        UsbStage.Ready => "USB", UsbStage.HighSpeed => Speed == UsbLinkSpeed.Unknown ? "USB" : "USB · High-Speed", UsbStage.FullSpeed => "USB · Full-Speed",
        UsbStage.Fallback => "Bluetooth RFCOMM",
        _ => BluetoothAvailable ? "Bluetooth RFCOMM" : Strings.Get("暂无可用通道"),
    };
    public bool IsReady => Stage is UsbStage.Ready or UsbStage.HighSpeed or UsbStage.FullSpeed;
}

public sealed record UsbDevice(string InstanceId, string Name, string InterfacePath, string Driver,
    bool AccessoryMode, bool AndroidCandidate, uint ProblemCode = 0, bool PolicyBlocked = false);
public sealed class UsbFailure(UsbStage stage, string message, Exception? inner = null) : IOException(message, inner)
{
    public UsbStage Stage { get; } = stage;
}
